src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java

```diff
@@ -104,7 +105,11 @@ public void put(String name, byte[] key, byte[] value, @Nullable Duration ttl) {
 			if (shouldExpireWithin(ttl)) {
 				connection.set(key, value, Expiration.from(ttl.toMillis(), TimeUnit.MILLISECONDS), SetOption.upsert());
 			} else {
-				connection.set(key, value);
+				if (shouldExpireMaxIdleWithin(maxIdle)) {
+					connection.set(key, value, Expiration.from(maxIdle), SetOption.upsert());
+				} else {
+					connection.set(key, value);
+				}
```

Comment:
As @mp911de alluded to above, there is technically only 1 expiration policy in Redis for a given key: TTL.

Expiration (and specifically, TTL) is something that is already [handled](https://github.com/spring-projects/spring-data-redis/blob/3.0.8/src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java#L104-L105) in Spring Data Redis as of `2.7.x` (and beyond: `3.0.x`, `3.1.x`, and currently, `3.2.x`), which is apparent in lines 105-107 shown above.

**Plus, what happens when TTL is less than TTI?** The lesser of TTL and TTI should take precedence.

Also note, as of Spring Data Redis `3.2.x` (currently in development with the latest version being `3.2.0-M1`) TTL expiration can even be computed dynamically, per cache entry by key (and value) using the new [TtlFunction](https://github.com/spring-projects/spring-data-redis/blob/3.2.0-M1/src/main/java/org/springframework/data/redis/cache/RedisCacheWriter.java#L174-L223) interface.

Reply:
Technically, TTI expiration policies (e.g. timeouts) have no bearing on cache entry creates or updates. TTL expiration is only updated when another create or update occurs after the initial create.

This would apply to `RedisCacheWriter.putIfAbsent(..)` as well.